pkg/cnpg/adapter.go

```diff
@@ -380,3 +536,122 @@ func Collectors(pool *pgxpool.Pool, kubeClient kubernetes.Client, podName string
+	if err != nil {
+		return nil, fmt.Errorf("failed to query pg_settings: %w", err)
+	}
+	defer rows.Close()
+
+	config := make(map[string]string)
+	for rows.Next() {
+		var name, setting string
+		var unit *string // Can be NULL
+		if err := rows.Scan(&name, &setting, &unit); err != nil {
+			return nil, fmt.Errorf("failed to scan pg_settings row: %w", err)
+		}
+
+		// Convert PostgreSQL format to CNPG format for comparison
+		// PostgreSQL returns: setting="262144", unit="8kB"
+		// We need: "2GB" (to match what CNPG expects)
+		if unit != nil && *unit != "" {
+			config[name] = ConvertWithUnit(setting, *unit)
+		} else {
+			config[name] = setting
+		}
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("error iterating pg_settings rows: %w", err)
+	}
+
+	return config, nil
+}
+
+// NormalizeValue normalizes parameter values for comparison.
+// Handles different representations: "2GB" == "2048MB" == "2097152kB"
+func NormalizeValue(value string) string {
+	normalized := strings.ToLower(strings.TrimSpace(value))
+
+	// Try to parse as memory value and convert to bytes
+	if bytes := parseMemoryToBytes(normalized); bytes > 0 {
+		return fmt.Sprintf("%d", bytes) // Return as bytes string
+	}
+
+	// For non-memory values, return normalized string
+	return normalized
+}
+
+// parseMemoryToBytes parses memory value and returns bytes.
+// Handles: "2GB", "256MB", "64kB", "1024" (plain number)
+func parseMemoryToBytes(value string) int64 {
+	value = strings.TrimSpace(value)
+
+	// Extract number and unit
+	var numStr string
+	var multiplier int64 = 1
+
+	if strings.HasSuffix(value, "gb") {
+		numStr = strings.TrimSpace(value[:len(value)-2])
+		multiplier = 1024 * 1024 * 1024
+	} else if strings.HasSuffix(value, "mb") {
+		numStr = strings.TrimSpace(value[:len(value)-2])
+		multiplier = 1024 * 1024
+	} else if strings.HasSuffix(value, "kb") {
+		numStr = strings.TrimSpace(value[:len(value)-2])
+		multiplier = 1024
+	} else {
+		numStr = value // Plain number
+	}
+
+	// Parse number (handles decimals like "2.5")
+	numValue, err := strconv.ParseFloat(numStr, 64)
+	if err != nil {
+		return 0 // Not a valid number
+	}
+
+	return int64(numValue * float64(multiplier))
+}
+
+// RequiresRestart checks if any of the changed parameters require database restart.
+// Returns true if any parameter has context='postmaster' in pg_settings.
+func (adapter *CNPGAdapter) RequiresRestart(ctx context.Context, changedParams map[string]string) (bool, error) {
+	if len(changedParams) == 0 {
+		return false, nil
+	}
+
+	// Build list of changed parameter names for query
+	paramNames := make([]string, 0, len(changedParams))
+	for name := range changedParams {
+		paramNames = append(paramNames, name)
+	}
+
+	// Query which of the changed parameters require restart
+	query := `SELECT name FROM pg_settings WHERE name = ANY($1) AND context = 'postmaster'`
+	rows, err := adapter.PGDriver.Query(ctx, query, paramNames)
```

Comment:
Use existing infrastructure, again so we escape it and handle it in a centralized place:
https://github.com/dbtuneai/dbtune-agent/blob/c28278857cf8d70e15dc4c1ea5de43cd9a2ae816/pkg/internal/utils/query_runtime.go#L21-L44